experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -89,112 +93,117 @@ LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
+      // A speculating branch first discards the condition in case that
+      // the data is not speculative. In case it is speculative, a new branch
+      // is created that replicates the current branch.
+
+      OpBuilder builder(ctx);
+      builder.setInsertionPointAfterValue(ctrlSignal);
+
+      // The speculating branch will discard the branch's condition token if
+      // the branch output is non-speculative. Speculative tag of the token is
+      // currently implicit, so the branch input itself is used at the IR
+      // level.
+      auto branchDiscardNonSpec =
+          builder.create<handshake::SpeculatingBranchOp>(
+              branchOp.getLoc(), valueForSpecTag /* specTag */,
+              branchOp.getConditionOperand());
+      inheritBB(specOp, branchDiscardNonSpec);
+
+      // The replicated branch directs the control token based on the path the
+      // speculative token took
+      auto branchReplicated = builder.create<handshake::ConditionalBranchOp>(
+          branchDiscardNonSpec->getLoc(),
+          branchDiscardNonSpec.getTrueResult() /* condition */,
+          ctrlSignal /* data */);
+      inheritBB(specOp, branchReplicated);
+
+      // Update ctrlSignal
+      ctrlSignal = branchReplicated->getResult(branchDirection);
+    }
     // Connect commit to the correct control signal and end traversal
     commitOp.setOperand(1, ctrlSignal);
   } else if (auto branchOp = dyn_cast<handshake::ConditionalBranchOp>(currOp)) {
-    // Replicate a branch in the control path and use new control signal.
-    // To do so, a structure of two connected branches is created.
-    // A speculating branch first discards the condition in case that
-    // the data is not speculative. In case it is speculative, a new branch
-    // is created that replicates the current branch.
-
-    MLIRContext *ctx = &getContext();
-    OpBuilder builder(ctx);
-    builder.setInsertionPointAfterValue(ctrlSignal);
-
-    // The speculating branch will discard the branch's condition token if the
-    // branch output is non-speculative. Speculative tag of the token is
-    // currently implicit, so the branch input itself is used at the IR level.
-    auto branchDiscardNonSpec = builder.create<handshake::SpeculatingBranchOp>(
-        branchOp.getLoc(), currOpOperand.get() /* specTag */,
-        branchOp.getConditionOperand());
-    inheritBB(specOp, branchDiscardNonSpec);
-
-    // The replicated branch directs the control token based on the path the
-    // speculative token took
-    auto branchReplicated = builder.create<handshake::ConditionalBranchOp>(
-        branchDiscardNonSpec->getLoc(),
-        branchDiscardNonSpec.getTrueResult() /* condition */,
-        ctrlSignal /* data */);
-    inheritBB(specOp, branchReplicated);
-
     // Follow the two branch results with a different control signal
     for (unsigned i = 0; i <= 1; ++i) {
       for (OpOperand &dstOpOperand : branchOp->getResult(i).getUses()) {
```

Comment:
I'm not sure we can assume materialization is performed before this pass. As I mentioned in another comment, we have good reasons for doing materialization afterward, so it's probably best to keep the implementation in a generalized way.

By the way, I like the way you zipped the results with the index - I'll give it a try!

Reply:
The point i wanted to make here is that getUses(0) works fine: getUses(1), getUses(2)… are added by yourself and I don’t think you want to traverse them when doing the DFS

Reply:
That's true... now I realize these cases occur when a loop has multiple BBs. A control signal gets routed for a commit, and then traversal continues into the newly routed speculating branch. But it doesn't cause a bug because the commit unit isn't visited again, which prevents multiple control networks from being routed for a single commit unit.

I want to confirm one thing: is it really okay to assume every pass runs immediately after materialization? If someone later decides it's inefficient and changes the flow, we might not remember this (though I'll leave a comment about it). And other parts of this pass also seem to handle the non-materialized form.

I also think there are other ways to prevent traversal from entering the commit control network.

Reply:
a situation like this?

```
    Op1
    / \
   |   |
Commit  SomethingElse
   |        |
   |        |
  Path_1   Path_2
```
(Commit and "SomethingElse" use the same value (channel))

we visit SomethingElse after Commit (https://github.com/EPFL-LAP/dynamatic/blob/d1442fd60c2e7271aed5b1b1f8510da773a0fe04/experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp#L193-L197

Assuming we know how to place commit units in unmaterialized IR, it looks like not all paths starting from "Op1" are guarded with commit though.

Reply:
I haven't read the specifics of whether this causes a bug or not, but I agree with Shun that I would rather not assume the circuit has been materialized if possible.

Given that we break materialize immediately, if another pass also breaks materialization in future it would be nice to be able to stack them without having calling materialize in between

Reply:
One materialization we do insist on is that whatever value we are speculating on has a single consumer- @shundroid we should probably add an explicit check for this at the beginning of the pass?

Reply:
I noticed that the number of BBs inside the loop doesn't matter. However, I found the traversal never explores the newly-created paths generated during this algorithm. (We can ignore other operations created in this pass but outside this algorithm.)

This situation is like Jiahui's case but is a more concrete example of it . In this context, SomethingElse corresponds to "replicating branches, etc."

![image](https://github.com/user-attachments/assets/5d93ca33-ce1d-4ab6-a8b4-5d27044f8266)

The figure above illustrates my updated understanding of why the traversal skips the "replicating branch."

However, I don't like this way. I previously discussed with Emmet a better way to handle commit routing, and I might change the timing of placing the actual units (eg. replicating branches).

> One materialization we do insist on is that whatever value we are speculating on has a single consumer
@murphe67 I think even this doesn't (ideally) assume materialization, since we specify the position using an operand. While value (OpResult) can be referenced by multiple operands without materialization, operands are always unique.
However, the current implementation seems to fail in the non-materialization case. My suggestion is to replace `replaceAllUsesExcept` with `operand->set` (just like the discussion in #168 ).

Reply:
@Jiahui17 Shun and I have come up with a more elegant solution to this problem, but we think it should be a full 2nd pull request. The current solution is 100% correct, it is just a little fragile.

Reply:
@shundroid for the materialization requirement, the speculator will need to direct the control token and at least one loop variable (unless the loop does the same thing every time). If the fork to direct the speculation does not exist yet, the algorithms will not fail, but speculation as a whole will not succeed.

Reply:
One terminology that I got a bit confused:
- The control token(s) sent to the discardNonSpec branch
- The control token(s) sent to inform speculation successful/wrong 

As far as what I got:
1. Commit routing thing: strictly splitting traversal and adding spec ctrl network would make the whole thing work without materialization.
2. Assuming no materialization: The speculator needs to send the value to both the spec ctrl network (the discardNonSpecBranch I assume?) and all its regular users.

Is the problem that the discardNonSpecBranch should receive something different from the other regular users?

Reply:
Oh the materialization requirement is nothing to do with the control network routing. 

Routing breaks materialization, but does not require materialization in the first place either.

The issue is if the condition we are speculating on has multiple unmaterialized uses, only one of them will ever receive speculative tokens.

Reply:
Now that I think about we could probably change the code to remove the assumption - it's an engineering side effect, not an intrinsic issue

Reply:
> The issue is if the condition we are speculating on has multiple unmaterialized uses, the way we place the speculator at the very beginning means only one of them will ever receive speculative tokens.

Now I see what Shun meant, you can just do replaceAllUses for this

Reply:
> The issue is if the condition we are speculating on has multiple unmaterialized uses, the way we place the speculator at the very beginning means only one of them will ever receive speculative tokens.

I think this behavior is "intended", but as Jiahui pointed out, now I realize it works in a somewhat "hacky" way in this unmaterialized case (though I wanted to mean another thing by mentioning replaceAllUses).

Currently, the placement JSON specifies the position based on the OpOperand, but the pass internally handles it as the Value (OpResult), because it utilizes replaceAllUsesExcept.

So possibly the place of the speculator should be correctly specified by the value (OpResult) in the file too, not by the operand. It supports this unmaterialized case "legally".
But in turn, it would prevent the case where we want to place the speculator on only one consumer of the value (though I’m not sure if such cases actually occur).

I now think it’s better to specify the position by OpOperand, which indirectly requires the materialization of at least the speculator position in most cases. Along with this, we replace `replaceAllUsesExcept` with `op->setOperand`, ensuring a more precise and consistent placement mechanism.